framework/src/org/checkerframework/common/signatureinference/SignatureInferenceScenes.java

```diff
@@ -0,0 +1,588 @@
+     * methodTree, then the type of the value passed to the return expression
+     * will be added to the return type of that method in the Scene.
+     * If the Scene previously contained an annotated return type for the
+     * method methodTree, its new type will be the LUB between the previous
+     * type and the type of the value passed to the return expression.
+     *
+     * @param retNode the node that contains the expression returned.
+     * @param classSymbol the symbol of the class that contains the method.
+     * @param atf the annotated type factory of a given type system, whose
+     * type hierarchy will be used to update the method's return type.
+     */
+    public static void updateMethodReturnTypeInScene(ReturnNode retNode,
+            ClassSymbol classSymbol, MethodTree methodTree,
+            AnnotatedTypeFactory atf) {
+        if (classSymbol == null) return;
+        AnnotatedTypeMirror methodReturnType = atf.getAnnotatedType(methodTree).
+                getReturnType();
+        if (methodReturnType.getExplicitAnnotations().size() > 0) {
+            // We do not infer types if there are explicit annotations.
+            // See https://github.com/typetools/annotation-tools/issues/105
+            return;
+        }
+        String jaifPath = JAIF_FILES_PATH + classSymbol.flatname.toString() +
+                ".jaif";
+        AScene scene = getScene(jaifPath);
+        AnnotatedTypeMirror returnExprATM = atf.getAnnotatedType(retNode.
+                getTree().getExpression());
+
+        AClass clazz = getJaifClass(classSymbol, scene);
+        if (clazz == null) return; // Anonymous class => Ignore, for now.
+        String methodName = JVMNames.getJVMMethodName(methodTree);
+        AMethod method = getJaifMethod(clazz, methodName, scene);
+        if (method.returnType != null) {
+            updateAnnotationSetInScene(
+                    method.returnType.tlAnnotationsHere, atf,
+                    jaifPath, returnExprATM);
+        }
+    }
+
+    /**
+     * Updates the set of annotations in a location of a Scene. If the set of
+     * annotations (curAnnos) is empty, then the updated set will be the set of
+     * annotations in newATM.
+     * If curAnnos is not empty, the updated set will be the LUB between
+     * curAnnos and newATM.
+     *
+     * @param curAnnos set of annotations located somewhere in a Scene.
+     * @param atf the annotated type factory of a given type system, whose
+     * type hierarchy will be used to update curAnnos.
+     * @param jaifPath identifies a Scene.
+     * @param newATM the type containing a set of annotations to be
+     * added (or "lubbed") to the Scene.
+     */
+    private static void updateAnnotationSetInScene(Set<Annotation> curAnnos,
+            AnnotatedTypeFactory atf, String jaifPath,
+            AnnotatedTypeMirror newATM) {
+        if (curAnnos != null && curAnnos.size() > 0) {
```

Comment:
This test suggests that curAnnos can be null, but curAnnos is not changed in the body and methods are called on it unconditionally later in the method.  So it's misleading to test against null.
If the contract of the method is that curAnnos is non-null, then you can add a separate test for that or you can just let the curAnnos.size() call fail.